### PR TITLE
Include some recent pitfalls in contributing.md

### PR DESCRIPTION
**Pull request by koivunej**
A non-exhaustive list.

Per our discussions with @LizardWizzard.

### CODE REVIEW

**Review by LizardWizzard — APPROVED** [review]
Nice!

**Review by LizardWizzard — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -9,6 +9,32 @@ refactoring, additional comments, and so forth. Let's try to raise the
...
+    Any future can be cancelled at every await point unless it's a top
+    level spawned future. Most common reasons to be cancelled or
+    dropped is that a future is polled or raced within `tokio::select!`
+    with a cancellation token or the HTTP client drops the connection,
+    causing a cancellation or drop on a handler future.
+    
+    Additionally please note that blocking tasks spawned with
+    `tokio::spawn_blocking` directly or indirectly will not be
+    cancelled with the spawning future.
+
```
Comment (hlinnaka) [inline CONTRIBUTING.md:27]: I'd like to add here that you are not expected to write async tasks in 100% cancellation safe manner. The cognitive burden of that is too high, and it's too easy to screw up.

Instead, the policy is to try to run all tasks to completion, unless you are very sure that the cancelled task is indeed cancellation safe. Dropping a tokio::sleep Future is fine for example, but cancelling anything more complicated is a red flag. Instead, use a CancellationToken to signal the Future to exit itself gracefully.

Comment (hlinnaka) [reply CONTRIBUTING.md:27]: #4194 is not related to async cancellation, is it?

Comment (koivunej) [reply CONTRIBUTING.md:27]: It is in the sense that it gives the naturally canceallable http handlers a way out to start operation of which at most one will run concurrently compared to spawning a task on each request. On top of not spawning more than one attempt or retry at a time, multiple http request handlers can wait for that one unique attempt to complete, and get the result.

If cancellation is wanted, it needs to be added to the one task manually (using `task_mgr::shutdown_token` for example).

Comment (hlinnaka) [reply CONTRIBUTING.md:27]: To summarize, we have two options:

a) Make all code cancellation safe. Inspect every `.await` carefully for safety, keeping in mind that the `.await` might never return.

b) Ensure that tasks are run to completion. Inspect every 'tokio::select!` carefully, to make sure that futures are not dropped. The only exception are futures that are known to be cancellation-safe, which covers things like tokio::time::sleep(), but nothing more complicated than that. Use CancellationTokens to cancel tasks gracefully.

I think a) is madness. Writing anything non-trivial in a cancellation-safe manner is error-prone.

Option b) is quite doable. `tokio::select!` and the few other functions that can drop Futures are pretty easy to grep, and we don't have very many of them in our codebase.

As it is, this text doesn't take a clear stance on this. It doesn't clearly say whether the code you write needs to be cancellation safe, or whether you need to run tasks to completion. It maybe implies a), maybe, but it's not clear. If we talk about async cancellation, I think we need to make a clear decision and call it out explicitly.

Btw, this should perhaps go to https://github.com/neondatabase/neon/blob/main/docs/pageserver-thread-mgmt.md, perhaps with a link from here.

Comment (hlinnaka) [reply CONTRIBUTING.md:27]: > It is in the sense that it gives the naturally canceallable http handlers a way out to start operation of which at most one will run concurrently compared to spawning a task on each request. 

Does hyper ever drop the Future that's handling an HTTP request?

If it does, then I think we need a wholesale solution, and e.g. spawn a new task for every request.

Comment (koivunej) [reply CONTRIBUTING.md:27]: > > It is in the sense that it gives the naturally canceallable http handlers a way out to start operation of which at most one will run concurrently compared to spawning a task on each request.
> 
> Does hyper ever drop the Future that's handling an HTTP request?

Yes, at least with HTTP/1.1 the way to cancel a request is to drop the connection. I am not aware of how it works with HTTP2. Of course it's nothing certain, like the connection going away could be noticed as late as finishing writing the response.

As discussed, I object to whole sale approach trying to work away with unconditional spawning which will still leave us with competing accesses and inability to know when the request has been in fact cancelled. #4194 solves both while being **explicit** in code about it. Spawning done somewhere far and not visible would be surprising. Async cancellation is a feature, we should rather embrace it than trying to close our eyes of it.

**Review by problame — APPROVED** [review]
I won't get into the remaining [open discussion](https://github.com/neondatabase/neon/pull/4198#discussion_r1190209316) but what's currently written here is correct and noteworthy.

**Pull request closed without merging**